w3af/plugins/grep/vulners_scanner.py

```diff
@@ -0,0 +1,219 @@
+        # That's only my opinions on the severity vs CVSS.score
+        # Maybe move it to the options later?
+        cvss_score = cvss_score * 10
+        if cvss_score in range(20, 30):
+            return severity.LOW
+        elif cvss_score in range(30, 70):
+            return severity.MEDIUM
+        elif cvss_score in range(70, 100):
+            return severity.HIGH
+        else:
+            return severity.INFORMATION
+
+    def grep(self, request, response):
+        """
+        Plugin entry point, search for vulnerable software banners  in web application HTTP responses.
+        :param request: The HTTP request object.
+        :param response: The HTTP response object
+        :return: None
+
+        """
+        # Lazy update rules if it's first start of the plugin
+        if not self.rules_updated:
+            self.update_vulners_rules()
+            self.rules_updated = True
+
+        # Check if we have downloaded rules well.
+        # If there is no rules - something went wrong, time to exit.
+        if not self.rules_table:
+            return
+        # We do not parse non-text output
+        if not response.is_text_or_html():
+            return
+        if response.get_url().get_domain_path() in self._already_visited:
+            return
+
+        self._already_visited.add(response.get_url().get_domain_path())
+
+        raw_response = response.dump()
+
+        # Here we will store uniq vulnerability map
+        vulnerabilities_summary = {}
+
+        for match, _, regex_comp, software_list in self._multi_re.query(raw_response):
+            # If RE matched we have
+            detected_version = match.group(1)
+            for software_name in software_list:
+                mathced_rule = self.rules_table[software_name]
+                vulnerabilities_map = self.check_vulners(software_name= mathced_rule['alias'].encode(),
+                                                         software_version= detected_version,
+                                                         check_type= mathced_rule['type'].encode())
+                flattened_vulnerability_list = [item for sublist in vulnerabilities_map.values() for item in sublist]
+                for bulletin in flattened_vulnerability_list:
+                    if bulletin['id'] not in vulnerabilities_summary:
+                        vulnerabilities_summary[bulletin['id']] = bulletin
+
+        # Now add KB's for found vulnerabilities
+        for bulletin in vulnerabilities_summary.values():
+
+            v = Vuln(name= bulletin['id'],
+                     desc = bulletin['description'] or bulletin.get('sourceData', bulletin['title']),
+                     severity = self.get_severity(bulletin.get('cvss', {}).get('score', 0)),
+                     response_ids = response.id,
+                     plugin_name = self.get_name())
+
+            v.set_url(response.get_url())
+
+            self.kb_append_uniq(location_a = self,
```

Comment:
Let's say that the HTTP response header shows a `Server` which triggers a vulnerability detection. An old vulnerable web server is in use in this application. All HTTP responses generated by the application include this server header and thus trigger the vulnerability detection. Will the scan end with 293810 findings (one for each URL) for this vulnerability?

I haven't run a scan to confirm, but it looks like that.

I believe you guys need to research https://github.com/andresriancho/w3af/blob/master/w3af/core/data/kb/info_set.py and group by vulnerability. In other words, report an `InfoSet` vulnerability with title: `Old Vulnerable Server` and then add `Info` instances to that `InfoSet`

Reply:
Got it. Fixed.